### DISCUSSION TITLE
3 Container Solution: Versioning?

### DISCUSSION THREAD

**Discussion by dm17** [nginx-proxy/acme-companion#804] [General]
Should the 3 container solution:
https://github.com/JrCs/docker-letsencrypt-nginx-proxy-companion/wiki/Docker-Compose

Have each image versioned? It seems the nginx.tmpl file is versioned - and I just got the latest to attempt to remedy a problem trying to get www subdomain working.
Thanks.

**Comment by buchdag** [comment]
My general advice would be to use tagged versions of images whenever possible rather than plain `latest` (either pulled from Dockerhub or built from sources and manually tagged if the maintainer is not tagging new releases along code changes).

This case is a bit specific as most of the functionnality of `nginx-proxy` depends on the `nginx.tmpl` file. In a three containers setup, the versioning of this file is under your own responsibility.

Hope that answers your question.

**Comment by sgabe** [comment]
The `nginx-proxy` container has `docker-gen` bundled with `nginx` and serves general purposes. The 3 containers approach requires some fine-tuning of the `nginx.tmpl` file. The [template](https://github.com/jwilder/docker-gen/blob/master/templates/nginx.tmpl) shipped with the `docker-gen` container is just a 5 years old example and the `nginx-proxy` [template](https://github.com/nginx-proxy/nginx-proxy/blob/master/nginx.tmpl) might not work out of the box. However, the 3 containers approach allows you to use a custom template with [docker-gen](https://github.com/jwilder/docker-gen) and [nginx-unprivileged](https://hub.docker.com/r/nginxinc/nginx-unprivileged) instead of `nginx-proxy`.
